File: mobile/backend/app/repositories/stock_ledger_repository.py
"""
Stock ledger repository for Firestore operations.
"""

from typing import List, Optional, Dict, Any

from app.repositories.base_repository import BaseRepository


class StockLedgerRepository(BaseRepository):
    """Repository for StockLedger documents (immutable audit trail)."""

    def __init__(self):
        super().__init__("stock_ledger")

    async def get_by_product(self, product_id: str, limit: int = 100) -> List[Dict[str, Any]]:
        """Get ledger entries for a product."""
        return await self.query(
            filters=[("product_id", "==", product_id)],
            order_by="recorded_at",
            direction="DESCENDING",
            limit=limit,
        )

    async def get_by_location(
        self, x: int, y: int, z: int = 0, floor: int = 0, limit: int = 100
    ) -> List[Dict[str, Any]]:
        """Get ledger entries for a specific location."""
        return await self.query(
            filters=[
                ("x", "==", x),
                ("y", "==", y),
                ("z", "==", z),
                ("floor", "==", floor),
            ],
            limit=limit,
        )

    async def get_filtered(
        self,
        product_id: Optional[str] = None,
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """Get ledger entries with optional filters."""
        filters = []
        if product_id:
            filters.append(("product_id", "==", product_id))
        return await self.query(
            filters=filters if filters else None,
            order_by="recorded_at",
            direction="DESCENDING",
            limit=limit,
        )

    async def get_stock_at_location(
        self, x: int, y: int, z: int, floor: int, product_id: str
    ) -> int:
        """Calculate total stock for a product at a location by summing ledger entries."""
        entries = await self.query(filters=[
            ("x", "==", x),
            ("y", "==", y),
            ("z", "==", z),
            ("floor", "==", floor),
            ("product_id", "==", product_id),
        ])
        return sum(e.get("quantity", 0) for e in entries)

File: mobile/frontend/lib/data/repositories/stock_ledger_repository.dart
import 'package:frontend/data/models/stock_ledger.dart';
import 'package:frontend/data/enums.dart';
import 'base_repository.dart';

class StockLedgerRepository extends BaseRepository<StockLedger> {
  StockLedgerRepository() : super('stock_ledger');

  @override
  StockLedger fromMap(Map<String, dynamic> map) => StockLedger.fromMap(map);

  @override
  Map<String, dynamic> toMap(StockLedger entity) => entity.toMap();

  // ── Domain queries ────────────────────────────────────────

  /// Get ledger entries for a specific product.
  Future<List<StockLedger>> getByProduct(String productId) {
    return query(
      where: 'product_id = ?',
      whereArgs: [productId],
      orderBy: 'recorded_at DESC',
    );
  }

  /// Get ledger entries for a specific location.
  Future<List<StockLedger>> getByLocation(int x, int y, int z, int floor) {
    return query(
      where: 'x = ? AND y = ? AND z = ? AND floor = ?',
      whereArgs: [x, y, z, floor],
      orderBy: 'recorded_at DESC',
    );
  }

  /// Get entries linked to an operation.
  Future<List<StockLedger>> getByOperation(String operationId) {
    return query(
      where: 'operation_id = ?',
      whereArgs: [operationId],
      orderBy: 'recorded_at ASC',
    );
  }

  /// Get entries by operation type.
  Future<List<StockLedger>> getByOperationType(OperationType type) {
    return query(
      where: 'operation_type = ?',
      whereArgs: [type.value],
      orderBy: 'recorded_at DESC',
    );
  }

  /// Get recent movements.
  Future<List<StockLedger>> getRecent({int limit = 50}) {
    return getAll(orderBy: 'recorded_at DESC', limit: limit);
  }

  /// Get net stock for a product at a location.
  Future<int> getNetStock(String productId, int x, int y, int z, int floor) async {
    final db = await database;
    final result = await db.rawQuery(
      'SELECT COALESCE(SUM(quantity), 0) as net FROM stock_ledger WHERE product_id = ? AND x = ? AND y = ? AND z = ? AND floor = ? AND is_deleted = 0',
      [productId, x, y, z, floor],
    );
    return (result.first['net'] as int?) ?? 0;
  }

  /// Get total stock for a product across all locations.
  Future<int> getTotalStock(String productId) async {
    final db = await database;
    final result = await db.rawQuery(
      'SELECT COALESCE(SUM(quantity), 0) as total FROM stock_ledger WHERE product_id = ? AND is_deleted = 0',
      [productId],
    );
    return (result.first['total'] as int?) ?? 0;
  }

  /// Get stock summary by floor.
  Future<Map<int, int>> getStockByFloor(String productId) async {
    final db = await database;
    final results = await db.rawQuery(
      'SELECT floor, COALESCE(SUM(quantity), 0) as total FROM stock_ledger WHERE product_id = ? AND is_deleted = 0 GROUP BY floor',
      [productId],
    );
    final map = <int, int>{};
    for (final row in results) {
      map[row['floor'] as int] = row['total'] as int;
    }
    return map;
  }
}
